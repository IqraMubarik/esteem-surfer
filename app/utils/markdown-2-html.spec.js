/* eslint-disable */

import markDown2Html, { linkify, sanitizeNode } from './markdown-2-html';
import { getTestData } from '../../test/data';
import addressParser from './address-parser';

const fs = require('fs');
const path = require('path');

describe('Markdown to Html', () => {
  describe('Replace Author Names', () => {
    it('Should replace author names with link for given string', () => {
      let input = 'lorem ipsum @dolor sit amet';
      let expected =
        'lorem ipsum <a class="markdown-author-link" data-author="dolor">@dolor</a> sit amet';
      expect(linkify(input)).toBe(expected);

      input = '@lorem ipsum @dolor sit amet';
      expected =
        '<a class="markdown-author-link" data-author="lorem">@lorem</a> ipsum <a class="markdown-author-link" data-author="dolor">@dolor</a> sit amet';
      expect(linkify(input)).toBe(expected);

      input = '@lorem @ipsum @dolor sit amet';
      expected =
        '<a class="markdown-author-link" data-author="lorem">@lorem</a> <a class="markdown-author-link" data-author="ipsum">@ipsum</a> <a class="markdown-author-link" data-author="dolor">@dolor</a> sit amet';
      expect(linkify(input)).toBe(expected);

      input = '@lorem @ipsum @dolor \n @sit amet';
      expected =
        '<a class="markdown-author-link" data-author="lorem">@lorem</a> <a class="markdown-author-link" data-author="ipsum">@ipsum</a> <a class="markdown-author-link" data-author="dolor">@dolor</a> \n <a class="markdown-author-link" data-author="sit">@sit</a> amet';
      expect(linkify(input)).toBe(expected);

      input = '@lorem @ipsum @dolor \n @Sit amet';
      expected =
        '<a class="markdown-author-link" data-author="lorem">@lorem</a> <a class="markdown-author-link" data-author="ipsum">@ipsum</a> <a class="markdown-author-link" data-author="dolor">@dolor</a> \n <a class="markdown-author-link" data-author="sit">@Sit</a> amet';
      expect(linkify(input)).toBe(expected);
    });
  });

  describe('Replace Tags', () => {
    it('Should replace tags with link for given string', () => {
      let input = 'lorem ipsum #dolor sit amet';
      let expected =
        'lorem ipsum <a class="markdown-tag-link" data-tag="dolor">#dolor</a> sit amet';
      expect(linkify(input)).toBe(expected);

      input = '#lorem ipsum #dolor sit amet';
      expected =
        '<a class="markdown-tag-link" data-tag="lorem">#lorem</a> ipsum <a class="markdown-tag-link" data-tag="dolor">#dolor</a> sit amet';
      expect(linkify(input)).toBe(expected);

      input = '#lorem #ipsum #dolor sit amet';
      expected =
        '<a class="markdown-tag-link" data-tag="lorem">#lorem</a> <a class="markdown-tag-link" data-tag="ipsum">#ipsum</a> <a class="markdown-tag-link" data-tag="dolor">#dolor</a> sit amet';
      expect(linkify(input)).toBe(expected);

      input = '#lorem #ipsum #dolor \n #sit amet';
      expected =
        '<a class="markdown-tag-link" data-tag="lorem">#lorem</a> <a class="markdown-tag-link" data-tag="ipsum">#ipsum</a> <a class="markdown-tag-link" data-tag="dolor">#dolor</a> \n <a class="markdown-tag-link" data-tag="sit">#sit</a> amet';
      expect(linkify(input)).toBe(expected);

      input = '#lorem #ipsum #dolor \n #Sit amet';
      expected =
        '<a class="markdown-tag-link" data-tag="lorem">#lorem</a> <a class="markdown-tag-link" data-tag="ipsum">#ipsum</a> <a class="markdown-tag-link" data-tag="dolor">#dolor</a> \n <a class="markdown-tag-link" data-tag="sit">#Sit</a> amet';
      expect(linkify(input)).toBe(expected);

      input = 'you are #1';
      expected = 'you are #1';
      expect(linkify(input)).toBe(expected);

      input = 'you are #1 #steemit-promo';
      expected =
        'you are #1 <a class="markdown-tag-link" data-tag="steemit-promo">#steemit-promo</a>';
      expect(linkify(input)).toBe(expected);
    });
  });

  describe('Replace Both Mentions and Tags', () => {
    it('Should replace', () => {
      let input = 'lorem ipsum #dolor sit @amet';
      let expected =
        'lorem ipsum <a class="markdown-tag-link" data-tag="dolor">#dolor</a> sit <a class="markdown-author-link" data-author="amet">@amet</a>';
      expect(linkify(input)).toBe(expected);

      input = 'lorem ipsum @#dolor sit amet';
      expected = 'lorem ipsum @#dolor sit amet';
      expect(linkify(input)).toBe(expected);
    });
  });

  describe('Sanitize Node', () => {
    it('Should convert not allowed tag to span', () => {
      const input = `<script>document.getElementById('body').remove();</script>`;
      const expected = "<span>document.getElementById('body').remove();</span>";

      const el = document.createElement('div');
      el.innerHTML = input;
      const res = sanitizeNode(el);

      expect(res.innerHTML).toBe(expected);
    });

    it('Should convert not allowed tag to span safely for wrong markup', () => {
      const input = `<script>document.getElementById('body').remove();<script>`;
      const expected =
        "<span>document.getElementById('body').remove();&lt;script&gt;</span>";

      const el = document.createElement('div');
      el.innerHTML = input;
      const res = sanitizeNode(el);

      expect(res.innerHTML).toBe(expected);
    });

    it('Should remove not allowed attributes', () => {
      const input = `<a title="Foo" onclick="document.bar()">Click</a>`;
      const expected = '<a title="Foo">Click</a>';

      const el = document.createElement('div');
      el.innerHTML = input;
      const res = sanitizeNode(el);

      expect(res.innerHTML).toBe(expected);
    });

    it('Should remove javascript links', () => {
      const input = `<a title="Foo" href="javascript:void(0)">Click</a>`;
      const expected = '<a title="Foo">Click</a>';

      const el = document.createElement('div');
      el.innerHTML = input;
      const res = sanitizeNode(el);

      expect(res.innerHTML).toBe(expected);
    });
  });

  describe('Mark Down -> Html', () => {
    it('Should convert markdown to html', () => {
      expect(markDown2Html('foo').trim()).toBe('<p>foo</p>');
    });
  });

  describe('Mark Down -> Html', () => {
    it('Should catch images in table', () => {
      const data = getTestData(
        'steemitboard',
        'steemitboard-notify-dunsky-20181210t153450000z'
      );
      expect(markDown2Html(data.body)).toMatchSnapshot();
    });
  });

  describe('Mark Down -> Html', () => {
    it('Should add https prefix', () => {
      expect(markDown2Html('<a href="foo">foo</a>').trim()).toBe(
        '<p><a class="markdown-external-link" data-href="https://foo">foo</a></p>'
      );
    });
  });

  describe('Mark Down -> Html', () => {
    it('Should replace busy links properly', () => {
      const data = getTestData(
        'muratkbesiroglu',
        'sci-fi-novel-underground-city-part-13'
      );
      expect(markDown2Html(data.body)).toMatchSnapshot();
    });
  });

  describe('Mark Down -> Html Legacy', () => {
    const dataDir = `${__dirname}/../../test/data/legacy/markdown-2-html`;
    let files = fs.readdirSync(dataDir);

    for (let file of files) {
      if (file === '.DS_Store') {
        continue;
      }
      const fileContents = fs.readFileSync(path.join(dataDir, file), 'utf8');
      let data = JSON.parse(fileContents);

      const id = data['id'];
      const input = data['input'];
      const expected = data['result'];

      it('ID: ' + id, function() {
        expect(markDown2Html(input)).toBe(expected);
      });
    }
  });

  describe('Mark Down -> Html', () => {
    it('Should not convert markdown links', () => {
      const input =
        'lorem [this error](https://steemitimages.com/0x0/https://d1vof77qrk4l5q.cloudfront.net/img/5752638e6965247789bc20cef34727263aaa41e1.png) ipsum';
      expect(markDown2Html(input)).toMatchSnapshot();
    });
  });
});
