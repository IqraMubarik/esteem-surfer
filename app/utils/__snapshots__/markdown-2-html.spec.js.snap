// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown to Html Mark Down -> Html Should catch images in table 1`] = `
"<p>Congratulations <a class=\\"markdown-author-link\\" data-author=\\"dunsky\\">@dunsky</a>! You have completed the following achievement on the Steem blockchain and have been rewarded with new badge(s) :</p>
<table><tbody><tr><td><a data-href=\\"https://steemitimages.com/60x70/http://steemitboard.com/@dunsky/votes.png?201812101133\\" class=\\"markdown-img-link\\"><img src=\\"https://steemitimages.com/0x0/https://steemitimages.com/60x70/http://steemitboard.com/@dunsky/votes.png?201812101133\\"></a></td><td>You made more than 6000 upvotes. Your next target is to reach 7000 upvotes.</td></tr>
<tr><td><a data-href=\\"https://steemitimages.com/60x70/http://steemitboard.com/@dunsky/posts.png?201812101316\\" class=\\"markdown-img-link\\"><img src=\\"https://steemitimages.com/0x0/https://steemitimages.com/60x70/http://steemitboard.com/@dunsky/posts.png?201812101316\\"></a></td><td>You published more than 250 posts. Your next target is to reach 300 posts.</td></tr>
</tbody></table>
<p><sub><em><a class=\\"markdown-external-link\\" data-href=\\"https://steemitboard.com/@dunsky\\">Click here to view your Board of Honor</a></em></sub><br>
<sub><em>If you no longer want to receive notifications, reply to this comment with the word</em> <code>STOP</code></sub></p>
<p>To support your work, I also upvoted your post!</p>
<blockquote>
<p>Support <a class=\\"markdown-external-link\\" data-href=\\"https://steemit.com/@steemitboard\\">SteemitBoard's project</a>! <strong><a class=\\"markdown-external-link\\" data-href=\\"https://v2.steemconnect.com/sign/account-witness-vote?witness=steemitboard&amp;approve=1\\">Vote for its witness</a></strong> and <strong>get one more award</strong>!</p>
</blockquote>
"
`;

exports[`Markdown to Html Mark Down -> Html Should not convert markdown links 1`] = `
"<p>lorem <a class=\\"markdown-external-link\\" data-href=\\"https://steemitimages.com/0x0/https://d1vof77qrk4l5q.cloudfront.net/img/5752638e6965247789bc20cef34727263aaa41e1.png\\">this error</a> ipsum</p>
"
`;

exports[`Markdown to Html Mark Down -> Html Should replace busy links properly 1`] = `
"<p><img src=\\"https://steemitimages.com/0x0/https://ipfs.busy.org/ipfs/QmPziJCkx8w62UZCz1TNxAMsgf1uTnPZdVKAYjLBWvMzeD\\" alt=\\"image.png\\"></p>
<p><a class=\\"markdown-post-link\\" data-tag=\\"busy\\" data-author=\\"muratkbesiroglu\\" data-permlink=\\"sci-fi-novel-underground-city-part-1-2\\">Part 1 - 2</a><br>
<a class=\\"markdown-post-link\\" data-tag=\\"busy\\" data-author=\\"muratkbesiroglu\\" data-permlink=\\"sci-fi-novel-underground-city-part-3\\">Part 3</a><br>
<a class=\\"markdown-post-link\\" data-tag=\\"busy\\" data-author=\\"muratkbesiroglu\\" data-permlink=\\"sci-fi-novel-underground-city-part-4\\">Part 4</a><br>
<a class=\\"markdown-post-link\\" data-tag=\\"busy\\" data-author=\\"muratkbesiroglu\\" data-permlink=\\"sci-fi-novel-underground-city-part-5\\">Part 5</a><br>
<a class=\\"markdown-post-link\\" data-tag=\\"busy\\" data-author=\\"muratkbesiroglu\\" data-permlink=\\"sci-fi-novel-underground-city-part-6\\">Part 6</a><br>
<a class=\\"markdown-post-link\\" data-tag=\\"busy\\" data-author=\\"muratkbesiroglu\\" data-permlink=\\"sci-fi-novel-underground-city-part-7\\">Part 7</a><br>
<a class=\\"markdown-post-link\\" data-tag=\\"busy\\" data-author=\\"muratkbesiroglu\\" data-permlink=\\"sci-fi-novel-underground-city-part-8\\">Part 8</a><br>
<a class=\\"markdown-post-link\\" data-tag=\\"busy\\" data-author=\\"muratkbesiroglu\\" data-permlink=\\"sci-fi-novel-underground-city-part-9\\">Part 9</a><br>
<a class=\\"markdown-post-link\\" data-tag=\\"busy\\" data-author=\\"muratkbesiroglu\\" data-permlink=\\"sci-fi-novel-underground-city-part-10\\">Part 10</a><br>
<a class=\\"markdown-post-link\\" data-tag=\\"busy\\" data-author=\\"muratkbesiroglu\\" data-permlink=\\"sci-fi-novel-underground-city-part-11\\">Part 11</a><br>
<a class=\\"markdown-post-link\\" data-tag=\\"busy\\" data-author=\\"muratkbesiroglu\\" data-permlink=\\"sci-fi-novel-underground-city-part-12\\">Part 12</a></p>
<p><strong>Part 13</strong></p>
<p>Members of the Yerşehir caravan have surrendered their weapons to imperial soldiers at the entrance of Istanbul by the agreement provided earlier. The imperial soldiers arrested them immediately after the surrender of the weapons, and they entered the city with their hands chained.</p>
<p>“The Yerşehir administration will ask you the account of this disrespect,” said Tamra to the proud Commander Timor on the pony. Tamra was angry because an ordinary soldier pushed her.</p>
<p>\\"You should be thankful we didn't strip you naked and had you wandering the streets of the city,\\" said Commander Timor.</p>
<p>Doma turned to the commander and said, “That's not how we dealt with you. Still, I assume you're doing your duty. I ask you to inform your superiors as soon as possible.”</p>
<p>”You may be executed for armed entry into the Imperial lands,\\" said the commander.</p>
<p>“You like to dramatize things,” Akman said. Although Commander Timor was angry with Akman's happy face, he chose to keep the sarcastic look on his face.</p>
<p>When he was taken to the dungeon, Akman was watching the buildings on the road, the squares adorned with sculptures, statues, and facades reflecting the beauty of East Roman stone craftsmanship.</p>
<p>After passing the Hippodrome, which is one of the rare works of the ancient world, they entered a narrow tunnel and went down a ladder descending by turning. The soldiers locked them in a large gallery and returned. The water was dripping from the high ceiling of the dungeon, and the air they breathed was quite damp. Akman saw two rats with the size of cats running to their holes, terrified by the sound of footsteps.</p>
<p>”I don't think they're gonna keep us here for long,\\" Akman said.</p>
<p>Akman was wrong; they told stories to each other in the first few days while there was no authority to call and ask for them. One of the dungeon guards came once a day and left a cup of water and a disgusting wheat porridge in the cell. The amount of oatmeal brought by the guard was so small that it was even possible for them to starve if the situation went on like this.</p>
<p>Akman was cheerful as if he was at the banquet table, not at the dungeon. His joy was so bizarre that the team members began to think he was out of his mind. As the days passed, he was weakened by malnutrition; his clothes were stained, and his face was blackened. Still, he didn't lose anything from his joy. When asked how it can be so enjoyable, he said, “peace and light are within us. Just let's call and find out.\\"</p>
<p>Days later, an officer finally appeared at the door of the dungeon. Now, even the sparkle in the eyes of Akman has been lost, and Doma was unable to get up from where he was sleeping because of his illness. “The emperor will meet with the two representatives of your choice in two days,” said Chlorus, who brought plenty of food with him. Chlorus was handling his business with care. He brought water and ash to the cell to make sure they didn't smell bad, and he arranged a basin for their clothes. After washing, a barber would cut off their hair and beard. When they talk to the emperor, they should not extend the word and that they should give satisfying answers to the questions asked.</p>
<p>Akman and Doma were taken from prison by the guards to be brought to the emperor. They agreed to hold a reasonable discourse during the interview. They were ready to suffer the maltreatment for the well-being of their country. The Imperial Palace adjacent to the Hippodrome was quite magnificent. They couldn't retrieve their eyes from the mosaics on the floor, the tiles on the walls and the stained glass windows. They passed through the large gate of the palace and entered the inner courtyard, where they waited for a while and then came to the emperor's presence.</p>
<p>Since Doma's disease has not passed yet, Akman was holding his arm. As a matter of fact, he hasn't been able to speak; he agreed to join the meeting to show The Emperor what a terrible treatment they had.</p>
<p>As soon as Emperor Valens saw Doma, he noticed his discomfort and told the servants to bring him a chair to sit down.</p>
<p>“You wanted to see me,” said the Emperor, sitting on a plain throne compared to the splendor of the palace.</p>
<p>”In the name of my people, I share your sorrow,\\" said Akman.</p>
<p>The emperor looked at Akman with the eyes asking, \\"what is it about?”</p>
<p>“I think robots attack your stables,” Akman said. Doma, who saw the environment as blurry due to high fever, turned and looked at Akman with amazement.</p>
<p>”Few people know that yet, \\" said Emperor Valens. Akman heard the news from the two soldiers they encountered on the way to the dungeon; the soldiers thought others did not understand them because they were whispering. As the security of his country is concerned, Akman chose to transfer the incident to the emperor in a different manner.</p>
<p>\\"We came to Istanbul to warn you about this. When the imperial authorities said in the dungeon that we could finally come to your peace, I understood that the attack has occurred,” Akman said.</p>
<p>“You speak our language very well,” said the emperor.</p>
<p>“Master Doma is much better than me, I hoped he'd be able to talk, but I was wrong,\\" Akman said.</p>
<p>\\"I give you my regards and my gratitude for showing us the grace of acceptance. Because of my illness, I beg forgiveness for the imperfections that I can handle,” Doma said.</p>
<p>“I will ask without hesitation: what could we have done to draw the wrath of God upon us, and what should we do to be merciful to his forgiveness?”</p>
<p>Akman did not tell the emperor that events had begun with Ordin rising to Earth. He didn't want him to think they were in charge of the activities. Moreover, there was no evidence that the attacks on the stables in Yerşehir and Istanbul were related to Ordin's action. It was more than three months ago, and it could have been a coincidence.</p>
<p>“I do not think that what happened is related to any sin.”</p>
<p>“Foreigners come to our city with magical weapons, and ten days later, such an attack is taking place. Is that a coincidence?”</p>
<p>\\"I suspect that your scientists have already studied our weapons and have partially solved their operational principles. I assure you, our weapons work in accordance with scientific rules. I don't know why we're being treated so hard in your country.”</p>
<p>“Could it be because you closed your doors to our country with a unilateral decision you made two centuries ago? According to historical records, despite the insistence of our empire, we have not even been able to negotiate the matter.”</p>
<p>“It was a great travesty that half of our people died because of the plague. But that's not a valid excuse. I think the decision we made on time was wrong. During the past, we have been deprived of the benefits that we can achieve mutually. Given that the tunnel system makes wars almost impossible, it is not reasonable for us to continue such a mistake,\\" Akman said.</p>
<p>“I haven't received a satisfactory answer to my questions yet?\\" the emperor said, despite the tone of his voice, his eyes were warm to Akman.</p>
<p>\\"I have to admit I don't know the answer to your questions, even if it's embarrassing for a philosopher. We came to Istanbul to find the answer to this question together.”</p>
<p>\\"I want to learn the working principles and how to produce the weapons you bring with you. This is the only way we can initiate an equality-based relationship.”</p>
<p>\\"I am not authorized to make such a decision. But I don't think your offer will be rejected at a time when we need cooperation to defend our cities,\\" Akman said.</p>
<p>After the meeting with Emperor Valens, the Yerşehir delegation was removed from prison, their belongings were returned, and they were sent from Istanbul with a ceremony in accordance with diplomatic traditions.</p>
<p>Image Source: pixabay.com</p>
"
`;
